set.c

```diff
@@ -0,0 +1,3555 @@
+/* This implements sets using the same hash table implementation as in
+   st.c, but without a value for each hash entry.  This results in the
+   same basic performance characteristics as when using an st table,
+   but uses 1/3 less memory.
+   */
+
+#include "id.h"
+#include "internal.h"
+#include "internal/bits.h"
+#include "internal/hash.h"
+#include "internal/proc.h"
+#include "internal/sanitizers.h"
+#include "internal/symbol.h"
+#include "internal/variable.h"
+#include "ruby_assert.h"
+
+#include <stdio.h>
+#ifdef HAVE_STDLIB_H
+#include <stdlib.h>
+#endif
+#include <string.h>
+
+#ifndef SET_DEBUG
+#define SET_DEBUG 0
+#endif
+
+#if SET_DEBUG
+#include "internal/gc.h"
+#endif
+
+#ifdef __GNUC__
+#define PREFETCH(addr, write_p) __builtin_prefetch(addr, write_p)
+#define EXPECT(expr, val) __builtin_expect(expr, val)
+#define ATTRIBUTE_UNUSED  __attribute__((unused))
+#else
+#define PREFETCH(addr, write_p)
+#define EXPECT(expr, val) (expr)
+#define ATTRIBUTE_UNUSED
+#endif
+
+#if SIZEOF_LONG == SIZEOF_VOIDP
+typedef unsigned long set_data_t;
+#elif SIZEOF_LONG_LONG == SIZEOF_VOIDP
+typedef unsigned LONG_LONG set_data_t;
+#else
+# error ---->> set.c requires sizeof(void*) == sizeof(long) or sizeof(LONG_LONG) to be compiled. <<----
+#endif
+#define SET_DATA_T_DEFINED
+
+#ifndef CHAR_BIT
+# ifdef HAVE_LIMITS_H
+#  include <limits.h>
+# else
+#  define CHAR_BIT 8
+# endif
+#endif
+#ifndef _
+# define _(args) args
+#endif
+#ifndef ANYARGS
+# ifdef __cplusplus
+#   define ANYARGS ...
+# else
+#   define ANYARGS
+# endif
+#endif
+
+typedef struct set_table set_table;
```

Comment:
It would be good to split the `set_table` internal implementation and the `Set` type, because there are other places in the VM that could benefit from using a `set_table` instead of a `st_table`, one example being the `fstring_table`.

Reply:
That makes sense.  I'll work on that.